### PR TITLE
Feature/member persistance

### LINKED ISSUE #31: [Bug] - While Refresh The Page Participants Left Room
### Bug Report

- [x]  I agree to follow this project's Contributing Guidelines.

**Is your feature request related to a problem? Please describe:**
When I refresh the page on Slate.AI, I am unexpectedly leaving the room. This behavior is not anticipated, and it disrupts the user experience.

**Describe the Solution You'd Like:**
I would like the system to remember the current room a user is in and retain that information even after a page refresh. The user should not be automatically removed from the room upon refreshing the page.

**Describe Alternatives You've Considered:**
An alternative solution could be implementing a session persistence mechanism that stores the user's current room information, allowing them to rejoin the room after a page refresh.

### PR DESCRIPTION
**Pull request by JavidSumra**
# Title and Issue number 

Title : While Refresh The Page Participants Left Room

Issue No. : #31 

Close #31 

# Video/Screenshots (mandatory)

https://github.com/sd012gfhkhhvh/Slate.AI/assets/112365664/579203f1-7461-48d5-9716-a5120296be5e

# Checklist:

- [x] I have mentioned the issue number in my Pull Request.
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have gone through the  `contributing.md` file before contributing

**Additional context (Mandatory )**
I am Using LocalStorage for User Persistence While Refresh the Page

***Are you contributing under any Open-source programme?***
I am Contributing Under JWOC 2024

### CODE REVIEW

**Comment by JavidSumra** [comment]
Hey @sd012gfhkhhvh Sir Please Have a Look at My PR and Don't Forgot to add `Labels`.🤝

File: server/server.js [hunk 1]
```diff
@@ -40,7 +40,7 @@ io.on("connection", (socket) => {
 
     console.log(`${userId} joined room ${roomId}`);
 
-    io.sockets.in(roomId).emit("userIsJoined", { users: users });
+    io.sockets.in(roomId).emit("userIsJoined", { users: data, success: true });
```
Comment (sd012gfhkhhvh) [inline server/server.js]: These line of changes crashing while clicking the button "peoples" in the room page. We have to emit the data of all the users in the room to get the peoples list on each window.

**Comment by sd012gfhkhhvh** [comment]
the issue is still there. after refrashing the socket connection got disconneted and a new socket connection got established and a new user joined as an undefined user with a undefined room ID. And also the undefined user gets all the data emitted by the room(This is the same bug as the chatbox issue, all the data of drawing elements are emmited to the users of other room. Would you like to fix that issue ?).

Screenshot 2024-01-27 142909

carefully observe the logs in the server.js, specially the last 3 lines. 

I am upgrading the issue with the level hard. try to debug these issues.

Suggested approach - make the user leave the room and redirect them in the join room page rather that joining an undefined user to the undefined room. and if possible give them a popup of rejoin with the same room credentials(room id and name) as they are in before.(not mendatory but appreciable if you do so)

**Comment by JavidSumra** [comment]
Yes sir i am interested to work on it @sd012gfhkhhvh Sir

**Comment by JavidSumra** [comment]
Hey @sd012gfhkhhvh Sir Done With Changes Now it Persist With Same Credentials and Chatbox is Fixed Now Please Check Once and Don't Forgot to Change Level.

File: client/src/pages/RoomPage/index.jsx [hunk 2]
```diff
@@ -34,15 +34,16 @@ export const RoomPage = ({ user, socket }) => {
   useEffect(() => {
     // add users in a array(usersInRoom)
     socket.on("userIsJoined", ({ users }) => {
-      console.log(users);
       setUsersInRoom(users);
```
Comment (sd012gfhkhhvh) [inline client/src/pages/RoomPage/index.jsx]: users suppose to be an array of users present in a perticular room, but it is not an array, it sets the current user to the state "usersInRoom" !!!

File: client/src/pages/RoomPage/index.jsx [hunk 3]
```diff
@@ -51,11 +52,15 @@ export const RoomPage = ({ user, socket }) => {
...
-      setUsersInRoom((prevUser) => {
-        return prevUser.filter((user) => user.socketId !== socketId);
+      if (name) {
+        toast.info(`${name} has left the room.`, {
+          position: toast.POSITION.TOP_CENTER,
+        });
+      }
+      setUsersInRoom(() => {
+        if (usersInRoom.length > 0) {
+          return usersInRoom?.filter((user) => user?.socketId !== socketId);
```
Comment (sd012gfhkhhvh) [inline client/src/pages/RoomPage/index.jsx]: if usersInRoom is not an array so on what you are filtering??

File: server/server.js [hunk 4]
```diff
@@ -40,13 +40,17 @@ io.on("connection", (socket) => {
...
```
Comment (sd012gfhkhhvh) [inline server/server.js]: the data you are sending is teh data of a single user not the users in the room.

File: server/server.js [hunk 5]
```diff
@@ -40,13 +40,17 @@ io.on("connection", (socket) => {
...
 
     socket
       .to(roomId)
       .emit("userJoinedRoom", { success: true, user: { socketId, ...data } });
   });
 
+  socket.on("getUserData", (data) => {
```
Comment (sd012gfhkhhvh) [inline server/server.js]: there is no need to create this extra socket event for users in a room, we are getting it under usersIsJoined event as soon as user joins the room.

**Comment by sd012gfhkhhvh** [comment]
@JavidSumra you have changed a lot more things which has no need. just handle the onDisconnect event correctly and redirect them to join room page with their credentials for rejoin. Dont make unnessary changes. Even you dont need useParms here as you can get all the data related to a particular disconnected userthrougn "user" state in the Room page. 

try not to make it complicated and messy using extra things that we donnt need at all

**Comment by sd012gfhkhhvh** [comment]
Did you test it properly in your local system? I have still noticed some inconsistencies. firstly after rejoin the user is only able to recieve data from the others in the room but cant emit data to others in the room. secondly there is an inconsistency of removing data from localstorage, after closing the modal the data does not get removed from the browser, thus when I try to create new room it pops up again

**Comment by JavidSumra** [comment]
Hey @sd012gfhkhhvh Sir it's Working for me but I will check ✅

**Comment by JavidSumra** [comment]
@sd012gfhkhhvh Sir, please review the PR. I've implemented the changes as per your instructions.

**Comment by sd012gfhkhhvh** [comment]
> Did you test it properly in your local system? I have still noticed some inconsistencies. firstly after rejoin the user is only able to recieve data from the others in the room but cant emit data to others in the room. secondly there is an inconsistency of removing data from localstorage, after closing the modal the data does not get removed from the browser, thus when I try to create new room it pops up again

@JavidSumra  the first point is stll not been resolved yet.

**Comment by JavidSumra** [comment]
Hey @sd012gfhkhhvh Sir Changes Done

**Comment by sd012gfhkhhvh** [comment]
> Hey @sd012gfhkhhvh Sir Changes Done

The eraser tool after rejoin is not working properly. 

Bro u are not testing the app properly before making a pr. It has been  3rd or 4th time i am saying this to you. when you are fixing any bug make sure to test properly before making aggressive pr. Its been very difficult for an open source project mentainer to invest their time again an again to a same pr for the same issue with repeated bugs.

**Comment by JavidSumra** [comment]
Hey @sd012gfhkhhvh Sir Currently I am busy with some college related stuff so sorry for this behaviour.

**Comment by sd012gfhkhhvh** [comment]
> Hey @sd012gfhkhhvh Sir Currently I am busy with some college related stuff so sorry for this behaviour.

Did both the participant's eraser tool works properly after one rejoins in your system? Send me a screen recording whenever you are free.

**Comment by sd012gfhkhhvh** [comment]
issue #31 has been resolved and has been merged under pr #32

**Pull request merged**